libherokubuildpack/src/write.rs

```diff
@@ -0,0 +1,212 @@
+use std::io;
+use std::mem;
+use std::sync::Arc;
+
+/// Constructs a writer that buffers written data until given marker byte is encountered and
+/// then applies a given mapping function to the data before passing the result to the wrapped
+/// writer.
+///
+/// See the [`mappers`] module for a collection of commonly used mappers.
+pub fn mapped<W: io::Write, F: (Fn(Vec<u8>) -> Vec<u8>) + Sync + Send + 'static>(
+    w: W,
+    marker_byte: u8,
+    f: F,
+) -> MappedWrite<W> {
+    MappedWrite {
+        inner: w,
+        marker_byte,
+        buffer: vec![],
+        mapping_fn: Arc::new(f),
+    }
+}
+
+/// Constructs a writer that buffers written data until an ASCII/UTF-8 newline byte (`0x0A`) is
+/// encountered and then applies a given mapping function to the data before passing the result to
+/// the wrapped writer.
+///
+/// See the [`mappers`] module for a collection of commonly used mappers.
+pub fn line_mapped<W: io::Write, F: (Fn(Vec<u8>) -> Vec<u8>) + Sync + Send + 'static>(
+    w: W,
+    f: F,
+) -> MappedWrite<W> {
+    mapped(w, NEWLINE_ASCII_BYTE, f)
+}
+
+/// Constructs a writer that writes to two other writers. Similar to the UNIX `tee` command.
+pub fn tee<A: io::Write, B: io::Write>(a: A, b: B) -> TeeWrite<A, B> {
+    TeeWrite {
+        inner_a: a,
+        inner_b: b,
+    }
+}
+
+/// A mapped writer that was created with the [`mapped`] or [`line_mapped`] function.
+pub struct MappedWrite<W: io::Write> {
+    inner: W,
+    marker_byte: u8,
+    buffer: Vec<u8>,
+    mapping_fn: Arc<dyn (Fn(Vec<u8>) -> Vec<u8>) + Sync + Send>,
+}
```

Comment:
I guess YAGNI is the way to go here - if there are no obvious benefits of implementing a specific trait (either manually or via `derive`) I would wait until we need it so we a) don't add things we don't need and b) know _how_ we should implement them. When thinking about `Debug` for example, what do we want, how should we represent the function? Do we need to?

Reply:
> So in general I think we should at least consider implementing these derives by default if we can: Clone, Default, PartialEq, PartialOrd, Hash, Eq, and Ord. Less because we'll directly need those things on this struct, and more that any struct that contains this one is blocked from deriving these traits if it contains our struct.

Some thoughts:
- In the general case, yes we should at least consider whether some of these are useful when adding new code.
- `and more that any struct that contains this one is blocked from deriving` -> good point, though in this specific case, why would they need to put the mapper in their own struct?
- Adding derives isn't free - it has an impact on compile times (there's been a bunch of performance work, eg [here](https://nnethercote.github.io/2022/07/20/how-to-speed-up-the-rust-compiler-in-july-2022.html), trying to reduce the size of the generated code, since in some cases it's quite verbose and gives the compiler extra work to do). This shouldn't stop us using them if there is a genuine chance someone will need them, but we should probably not default to adding all 7 of those derive attributes to every struct without thinking about it at least briefly case by case.

Reply:
Looking at the standard library for similar types that serve as intermediate types (often named after the function that creates them: `FlatMap`, `Child`, etc.), I changed my mind and think `Clone` and `Debug` make sense here. Except for `Debug` in the case of `MappedWrite`, they can be automatically derived and the `Debug` implementation isn't too bad.

I agree with the "lost momentum" argument. :)

Thanks for raising this! Added in 86462bee1558713fc4f757d1ad2240b4a6ce4248.

Reply:
Thank you for this @Malax ❤️ . 

> Adding derives isn't free

I meant in terms of programmer/maintainer time. But I hadn't really considered binary size or compile time. In my head I'm still basically laughing at how much faster even "bad" rust code is compared to carefully written Ruby code. It would be nice to get a sense of the cost of that tradoff (size and compile time). I'm not sure how we could do it.

Maybe we could remove deriving Debug from all structs in libcnb, clear the caches and time the compile time and check binary size? For me, i'm in a mode of usually wanting to optimize programmer productivity and experience first and then really only caring about perf and size second. I care a lot about performance in Ruby land mostly as a factor of it ultimately affecting productivity and experience. I'm curious to brainstorm ways to better tune intuition on these tradeoffs.

I think it's a good point, I want to know more. I always want to understand the impact of my (coding) actions.

> why would they need to put the mapper in their own struct?

More coming from a place of assuming people will build structs (and enums) from pretty much any other struct that exists. In this case, I've stored a bunch of stuff for command execution in a struct in my buildpack.

If there's a good reason to NOT allow someone to use or store a struct in their own code, we should probably annotate/document that case (or make the structs private all together). Otherwise as library implementers we need to write for what people might likely do, even if we've not got concrete ideas for why. In this case I don't think leaving out Clone or Debug would make someone second guess their code "hmm, maybe I shouldn't store this in a struct/enum" and instead they would likely jump straight to opening an issue or a PR. 

In general it's good to be opinionated and also have a good sense for what we want to encourage people to do. I think we also need to balance that with some general guidelines about reasonable expectations of what our users will and won't reasonably do.

Reply:
I'm not talking about not adding `Debug` (of all the traits, that seems like a no-brainer to add; my post was about the "7 traits" aspect, per: "we should probably not default to adding all 7 of those derive attributes to every struct without thinking about it at least briefly case by case"), nor am I talking about binary size or runtime performance (I only mention compile times). 

I know it's not intentional, but it feels like the response above is nearing straw man territory? :-)